glogger/handler.py

```diff
@@ -0,0 +1,277 @@
+
+    def _flush_loop(self) -> None:
+        self.last_flush_time = self.time_fn()
+        while not self.stop_event.is_set():
+            if self.should_flush():
+                self._flush()
+            self.stop_event.wait(0.5)
+        # Flush all remaining messages before terminating:
+        if self.messages_buffer.count > 0:
+            self._flush()
+
+    @property
+    def time_since_last_flush(self) -> float:
+        return self.time_fn() - self.last_flush_time
+
+    def should_flush(self) -> bool:
+        return (
+            self.messages_buffer.count > 0
+            and (self.messages_buffer.utilized >= self.flush_threshold)
+            or (self.time_since_last_flush >= self.flush_interval)
+        )
+
+    # This is deliberately not "flush", because logging.shutdown() calls flush() and we don't want
+    # any flushing to happen from multiple threads.
+    def _flush(self) -> None:
+        # Allow configuring max_tries via class member. Alternatively we could decorate _send_logs in __init__ but that
+        # would be easier to miss and less flexible.
+        send = backoff.on_exception(backoff.expo, exception=RequestException, max_tries=self.max_send_tries)(
+            self._send_logs
+        )
+
+        self.last_flush_time = self.time_fn()
+        try:
+            batch, response = send()
+            response.raise_for_status()
+        except Exception:
+            logger.exception(SERVER_SEND_ERROR_MESSAGE)
+        else:
+            self._drop_sent_batch(batch)
+
+    def _send_logs(self) -> Tuple[Batch, requests.Response]:
+        """
+        :return: (batch that was sent, response)
+        """
+        # Upon every retry we will remake the batch, in case we are able to batch more messages together.
+        batch = self.make_batch()
+        batch_data = {
+            "batch_id": batch.ident,
+            "metadata": self.get_metadata(),
+            "lost": batch.lost,
+            "logs": "<LOGS_JSON>",
+        }
+        # batch.logs is a list of json strings so ",".join() it into the final json string instead of json-ing the list.
+        data = self.jsonify(batch_data).replace('"<LOGS_JSON>"', f"[{','.join(batch.logs)}]").encode("utf-8")
+        # Default compression level (9) is slowest. Level 6 trades a bit of compression for speed.
+        data = gzip.compress(data, compresslevel=6)
+        response = self.session.post(
+            self.uri,
+            data=data,
+            headers={
+                "Content-Encoding": "gzip",
+                "Content-Type": "application/json",
+            },
+            timeout=self.request_timeout,
+        )
+        return batch, response
+
+    def _drop_sent_batch(self, sent_batch: Batch) -> None:
+        # we don't override createLock(), so lock is not None
+        with self.lock:  # type: ignore
+            # The previous lost count has been accounted by the server:
+            self.messages_buffer.dropped -= sent_batch.lost
```

Comment:
There is some discrepency in naming - dropped vs lost - let's please choose one - I suggested a naming in another comment.

Reply:
TODO: add doc about terminology